File: adapters/ios/StarbidAdMob/Sources/StarbidBannerCustomEvent.swift
import Foundation
import UIKit
import WebKit
import StarbidCore

public class StarbidBannerCustomEvent: NSObject, GADCustomEventBanner {

    public weak var delegate: GADCustomEventBannerDelegate?

    private var webView: WKWebView?
    private var containerView: UIView?
    private var ad: StarbidAd?

    public override init() {
        super.init()
    }

    public func requestAd(_ adSize: GADAdSize, parameter serverParameter: String?, label serverLabel: String?, request: GADCustomEventRequest) {

        let params = StarbidCustomEvent.parseServerParameters(serverParameter)
        StarbidCustomEvent.initializeSDK(params: params)

        guard let placementId = params[StarbidCustomEvent.paramPlacementId], !placementId.isEmpty else {
            let error = NSError(
                domain: "io.starbidz.admob",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Missing placement_id in server parameters"]
            )
            delegate?.customEventBanner(self, didFailAd: error)
            return
        }

        let width = Int(adSize.width)
        let height = Int(adSize.height)

        Starbidz.shared.requestBid(
            placementId: placementId,
            format: .banner,
            width: width,
            height: height
        ) { [weak self] result in
            guard let self = self else { return }

            DispatchQueue.main.async {
                switch result {
                case .success(let ad):
                    self.ad = ad
                    self.createBannerView(ad: ad, width: width, height: height)

                case .error(let message, _):
                    let error = NSError(
                        domain: "io.starbidz.admob",
                        code: 2,
                        userInfo: [NSLocalizedDescriptionKey: message]
                    )
                    self.delegate?.customEventBanner(self, didFailAd: error)

                case .noBid:
                    let error = NSError(
                        domain: "io.starbidz.admob",
                        code: 204,
                        userInfo: [NSLocalizedDescriptionKey: "No fill"]
                    )
                    self.delegate?.customEventBanner(self, didFailAd: error)
                }
            }
        }
    }

    private func createBannerView(ad: StarbidAd, width: Int, height: Int) {
        containerView = UIView(frame: CGRect(x: 0, y: 0, width: width, height: height))
        containerView?.backgroundColor = .clear

        let config = WKWebViewConfiguration()
        config.allowsInlineMediaPlayback = true

        webView = WKWebView(frame: containerView!.bounds, configuration: config)
        webView?.navigationDelegate = self
        webView?.scrollView.isScrollEnabled = false
        webView?.isOpaque = false
        webView?.backgroundColor = .clear
        webView?.autoresizingMask = [.flexibleWidth, .flexibleHeight]

        containerView?.addSubview(webView!)

        switch ad.creative.type {
        case .html:
            let html = wrapHtml(ad.creative.content)
            webView?.loadHTMLString(html, baseURL: nil)

        case .image:
            let imgHtml = """
            <html>
            <body style="margin:0;padding:0;">
            <img src="\(ad.creative.content)" style="width:100%;height:100%;object-fit:contain;"/>
            </body>
            </html>
            """
            webView?.loadHTMLString(imgHtml, baseURL: nil)

        case .vast:
            let error = NSError(
                domain: "io.starbidz.admob",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "VAST not supported for banner"]
            )
            delegate?.customEventBanner(self, didFailAd: error)
        }
    }

    private func wrapHtml(_ content: String) -> String {
        return """
        <!DOCTYPE html>
        <html>
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
            <style>
                * { margin: 0; padding: 0; box-sizing: border-box; }
                html, body { width: 100%; height: 100%; overflow: hidden; }
            </style>
        </head>
        <body>
            \(content)
        </body>
        </html>
        """
    }
}

// MARK: - WKNavigationDelegate

extension StarbidBannerCustomEvent: WKNavigationDelegate {

    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        guard let container = containerView else { return }
        delegate?.customEventBanner(self, didReceiveAd: container)

        if let ad = ad {
            Starbidz.shared.trackImpression(bidId: ad.bidId, placementId: "")
        }
    }

    public func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        delegate?.customEventBanner(self, didFailAd: error)
    }

    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {

        if navigationAction.navigationType == .linkActivated {
            delegate?.customEventBannerWasClicked(self)

            if let ad = ad {
                Starbidz.shared.trackClick(bidId: ad.bidId, placementId: "")
            }

            if let url = navigationAction.request.url {
                UIApplication.shared.open(url)
            }

            decisionHandler(.cancel)
            return
        }

        decisionHandler(.allow)
    }
}

File: adapters/android/starbidz-admob/src/main/kotlin/io/starbidz/admob/StarbidBannerCustomEvent.kt
package io.starbidz.admob

import android.content.Context
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.mediation.MediationAdLoadCallback
import com.google.android.gms.ads.mediation.MediationBannerAd
import com.google.android.gms.ads.mediation.MediationBannerAdCallback
import com.google.android.gms.ads.mediation.MediationBannerAdConfiguration
import io.starbidz.core.AdFormat
import io.starbidz.core.AdResult
import io.starbidz.core.StarbidzAd
import io.starbidz.core.Starbidz
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class StarbidBannerCustomEvent : MediationBannerAd {

    private var context: Context? = null
    private var adCallback: MediationBannerAdCallback? = null
    private var webView: WebView? = null
    private var container: FrameLayout? = null
    private var ad: StarbidzAd? = null

    fun loadAd(
        configuration: MediationBannerAdConfiguration,
        callback: MediationAdLoadCallback<MediationBannerAd, MediationBannerAdCallback>
    ) {
        context = configuration.context
        val serverParams = configuration.serverParameters
        val placementId = serverParams.getString(StarbidCustomEvent.PARAM_PLACEMENT_ID)

        if (placementId.isNullOrEmpty()) {
            callback.onFailure(AdError(1, "Missing placement_id", "io.starbidz"))
            return
        }

        val adSize = configuration.adSize
        val width = adSize.getWidthInPixels(context!!)
        val height = adSize.getHeightInPixels(context!!)

        CoroutineScope(Dispatchers.Main).launch {
            when (val result = Starbidz.requestBid(
                context!!,
                placementId,
                AdFormat.BANNER,
                adSize.width,
                adSize.height
            )) {
                is AdResult.Success -> {
                    ad = result.ad
                    createBannerView(result.ad, width, height)
                    adCallback = callback.onSuccess(this@StarbidBannerCustomEvent)
                }
                is AdResult.Error -> {
                    callback.onFailure(AdError(2, result.message, "io.starbidz"))
                }
                is AdResult.NoBid -> {
                    callback.onFailure(AdError(3, "No fill", "io.starbidz"))
                }
            }
        }
    }

    private fun createBannerView(ad: StarbidzAd, width: Int, height: Int) {
        val ctx = context ?: return

        container = FrameLayout(ctx).apply {
            layoutParams = FrameLayout.LayoutParams(width, height)
        }

        webView = WebView(ctx).apply {
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true

            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView?, url: String?) {
                    super.onPageFinished(view, url)
                    adCallback?.reportAdImpression()

                    CoroutineScope(Dispatchers.IO).launch {
                        Starbidz.trackImpression(ad.bidId, "")
                    }
                }
            }
        }

        container?.addView(webView)

        when (ad.creative.type) {
            StarbidzAd.CreativeType.HTML -> {
                webView?.loadDataWithBaseURL(null, wrapHtml(ad.creative.content), "text/html", "UTF-8", null)
            }
            StarbidzAd.CreativeType.IMAGE -> {
                val imgHtml = """
                    <html><body style="margin:0;padding:0;">
                    <img src="${ad.creative.content}" style="width:100%;height:100%;object-fit:contain;"/>
                    </body></html>
                """.trimIndent()
                webView?.loadDataWithBaseURL(null, imgHtml, "text/html", "UTF-8", null)
            }
            else -> {}
        }
    }

    override fun getView(): View = container!!

    private fun wrapHtml(content: String): String {
        return """
            <!DOCTYPE html>
            <html><head>
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <style>* { margin: 0; padding: 0; }</style>
            </head><body>$content</body></html>
        """.trimIndent()
    }
}
